docs/core/diagnostics/diagnosticsource-diagnosticlistener.md

```diff
@@ -0,0 +1,326 @@
+It then explains how <xref:System.Diagnostics.DiagnosticSource?displayProperty=nameWithType> consumes the data by first
+finding the specified DiagnosticListeners one is interested in, subscribing to them, and decoding payloads for more specific data.
+It finishes by describing filtering, which controls which events pass through the system.
+
+## Log an event
+
+The `DiagnosticSource` type is an abstract base class that defines the methods needed to log events. The class that holds the implementation is `DiagnosticListener`.
+The first step in instrumenting code with `DiagnosticSource` is to create a
+`DiagnosticListener`. For example:
+
+```C#
+    private static DiagnosticSource httpLogger = new DiagnosticListener("System.Net.Http");
+```
+
+Notice that httpLogger is typed as a `DiagnosticSource`.
+This is because this code
+only cares about writing events and thus only cares about the  `DiagnosticSource` methods that
+the `DiagnosticListener` implements. `DiagnosticListeners` are given names when they are created
+and this name should be the name of logical grouping of related events (typically the component).
+Later this name is used to find the Listener and subscribe to any of its events. `DiagnosticListeners` have a name, which is used to represent the component associated with the event.
+Thus the event names only need to be unique within a component.
+
+-------------------------------------------
+
+## Instrumenting with DiagnosticSource/DiagnosticListener
+
+We show how to instrument DiagnosticListener using the DiagnosticListener we made in the code above.
+
+The `DiagnosticSource` logging
+interface consists of two methods:
+
+```C#
+    bool IsEnabled(string name)
+    void Write(string name, object value);
+```
+
+A typical call site will look like:
+
+```C#
+    if (httpLogger.IsEnabled("RequestStart"))
+        httpLogger.Write("RequestStart", new { Url="http://clr", }); //any object can be the second argument
+```
+
+Architectural elements shown in the above code are as follows.
+Every event has a `string` name (e.g. `RequestStart`), and exactly one `object` as a payload.
+If you need to send more than one item, you can do so by creating an `object` with all information
+in it as properties. C#'s [anonymous type](https://docs.microsoft.com/dotnet/csharp/fundamentals/types/anonymous-types)
+feature is typically used to create a type to pass 'on the fly', and makes this scheme very
+convenient. At the instrumentation site, you must guard the call to `Write()` with an `IsEnabled()` check on
+the same event name. Without this check, even when the instrumentation is inactive, the rules
+of the C# language require all the work of creating the payload `object` and calling `Write()` to be
+done, even though nothing is actually listening for the data. By guarding the `Write()` call, we
+make it efficient when the source is not enabled.
+
```

Comment:
You might to introduce the code you are working with at the top of the section and then explain it as you go. To make this example a little more realistic we should probably change the naming. In practice a developer would be starting with some code that does some useful work and then they modify that code to add logging. The names of the class and method would still be based on the original work the class was doing.

```C#
// before adding logging
class HttpClient
{
    byte[] SendWebRequest(string url)
    {
        // pretend this sends an HTTP request to the url and gets back a reply
        byte[] reply = new byte[] {};
        return reply;
}
    
// after adding logging
class HttpClient
{
    private static DiagnosticSource logger= new DiagnosticListener("System.Net.Http");
    public byte[] SendWebRequest(string url)
    {
        if (logger.IsEnabled("RequestStart"))
        {
            httpLogger.Write("RequestStart", new { Url = url });
        }
        // pretend this sends an HTTP request to the url and gets back a reply
        byte[] reply = new byte[] {};
        return reply;
    }
}
```